### PR TITLE
History support

### PR DESCRIPTION
**Pull request by ajordens**
- Applications/Pipelines/Pipeline Strategies (to start)
- Only supported by S3 (Cassandra will return [])

### CODE REVIEW

**Comment by ajordens** [comment]
PTAL @spinnaker/netflix-reviewers 

A simple abstraction over going straight to an S3 bucket to look at historical pipeline/strategy/application configurations.

Currently just returns a full set of objects with timestamps ... figure that any `diff` should occur upstream.

File: front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pipeline/Pipeline.groovy [hunk 1]
```diff
@@ -22,9 +22,6 @@ import com.netflix.spinnaker.front50.model.Timestamped
 
 class Pipeline extends HashMap<String, Object> implements Timestamped {
   @JsonIgnore
-  Long lastModified
```
Comment (ajordens) [inline front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pipeline/Pipeline.groovy]: I don't believe having this actually go across the wire will cause grief but I'll give it a whirl in prestaging.

File: front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pipeline/PipelineDAO.groovy [hunk 2]
```diff
@@ -23,4 +23,5 @@ import com.netflix.spinnaker.front50.model.ItemDAO
 interface PipelineDAO extends ItemDAO<Pipeline> {
   String getPipelineId(String application, String pipelineName)
   Collection<Pipeline> getPipelinesByApplication(String application)
+  Collection<Pipeline> getPipelineHistory(String id)
```
Comment (cfieber) [inline front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pipeline/PipelineDAO.groovy]: should these be bounded somehow (max results param or something)?

File: front50-s3/src/main/java/com/netflix/spinnaker/front50/model/S3Support.java [hunk 3]
```diff
@@ -113,6 +113,30 @@ public T findById(String id) throws NotFoundException {
     }
   }
 
+  public Collection<T> allVersionsOf(String id) throws NotFoundException {
+    try {
+      VersionListing versionListing = amazonS3.listVersions(new ListVersionsRequest(bucket, buildS3Key(id), null, null, null, 20));
```
Comment (cfieber) [inline front50-s3/src/main/java/com/netflix/spinnaker/front50/model/S3Support.java]: ah - I see 20 maxResults here - maybe make that configurable (and / or support it as a request option)

File: front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pipeline/PipelineDAO.groovy (hunk 2 above)
Comment (ajordens) [inline front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pipeline/PipelineDAO.groovy]: Probably ... I just shoved in an explicit max of 20.

File: front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/StrategyController.groovy [hunk 4]
```diff
@@ -48,6 +48,15 @@ class StrategyController {
         pipelineStrategyDAO.getPipelinesByApplication(application)
     }
 
+    @RequestMapping(value = '{id:.+}/history', method = RequestMethod.GET)
+    Collection<Pipeline> getHistory(@PathVariable String id) {
+        try {
+            return pipelineStrategyDAO.getPipelineHistory(id)
+        } catch (UnsupportedOperationException ignored) {
+            return []
```
Comment (cfieber) [inline front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/StrategyController.groovy]: if the operation is supported, is the current version one of the values returned?

if it is, should the unsupported case just return the current value, or should the history request filter out the current value?

**Comment by cfieber** [comment]
LGTM

**Pull request merged**